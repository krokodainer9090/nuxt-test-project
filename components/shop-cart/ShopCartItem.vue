<script>
// Components
import Input from '../controls/input/TheInput';

export default {
  name: 'ShopCartItem',

  components: {
    Input
  },

  props: {
    cartItem: {
      type: Object,
      default: _ => ({})
    }
  },

  data: () => ({
    isError: false
  }),

  computed: {
    currentItemCount: {
      get () {
        return String(this.cartItem.currentCount);
      },

      set (currentCount) {
        if ( currentCount <= this.cartItem.count ) {
          this.isError = false;

          this.$emit('increase-count', {
            id: this.cartItem.id,
            count: currentCount
          });
        } else {
          this.isError = true;
        }
      }
    },

    itemName () {
      return this.cartItem?.N || '';
    },

    itemPrice () {
      return this.cartItem?.price
        ? `${this.cartItem.price} руб.`
        : '';
    }
  },

  methods: {
    deleteItem () {
      this.$emit('delete-item', this.cartItem.id);
    }
  }
};
</script>

<template>
  <div class="shop-cart-item">
    <div class="row shop-cart-item__content">
      <div class="col-6">
        <span
          v-if="itemName"
          class="shop-cart-item__name"
        >
          {{ itemName }}
        </span>
      </div>
      <div class="col-2">
        <Input
          v-model="currentItemCount"
          type="number"
          class="shop-cart-item__input"
        />
        <div
          v-if="isError"
          class="shop-cart-item__error"
        >
          <span class="shop-cart-item__error-text">
            Столько товара на складе нет. Осталось товара: {{ cartItem.count }} шт.
          </span>
        </div>
      </div>
      <div class="col-2 col-md-3">
        <span
          v-if="itemPrice"
          class="shop-cart-item__price"
        >
          {{ itemPrice }}
          <span class="shop-cart-item__price-count">/ шт.</span>
        </span>
      </div>
      <div class="col-2 col-md-1">
        <div
          class="shop-cart-item__delete-btn"
          @click="deleteItem"
        >
          <span class="shop-cart-item__delete-btn-text">
            Удалить
          </span>
        </div>
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../styles/components/shop-cart/shop-cart-item";
</style>
