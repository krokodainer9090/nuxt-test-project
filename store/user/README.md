В сторе user будет хранить состояние значения валюты и корзины пользователя.

С помощью action'ов addItemToCart, increaseItemCount, removeItemFromCart изменяем состояние корзины.

Action updateCartPrices отвечает за обновление цены в корзине пользователя (когда по интервалу в 15 секунд будем
получать данные из файла)
