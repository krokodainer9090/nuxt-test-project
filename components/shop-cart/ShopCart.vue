<script>
// Components
import ShopCartItem from './ShopCartItem';

// Vuex
import { createNamespacedHelpers } from 'vuex';

// Store
const user = createNamespacedHelpers( 'user' );

export default {
  name: 'ShopCart',

  components: {
    ShopCartItem
  },

  computed: {
    ...user.mapGetters({
      userCart: 'getUserCart',
      userCartSum: 'getUserCartSum'
    }),

    cartItems () {
      return this.userCart ?? [];
    }
  },

  methods: {
    getShopCartItemKey (key) {
      return `shop-card-item-${key}`;
    },

    increaseCountItem (item) {
      this.$store.dispatch('user/increaseItemCount', item);
    },

    deleteItem (itemId) {
      this.$store.dispatch('user/removeItemFromCart', itemId);
    }
  }
};
</script>

<template>
  <section class="shop-cart">
    <div class="container">
      <div class="shop-cart__content">
        <h2 class="shop-cart__title">Корзина</h2>
        <div class="shop-cart__header">
          <div class="row">
            <div class="col-6">
              <span class="shop-cart__header-text">
                Наименование товара и описание
              </span>
            </div>
            <div class="col-2">
              <span class="shop-cart__header-text">
                Количество
              </span>
            </div>
            <div class="col-4">
              <span class="shop-cart__header-text">
                Цена
              </span>
            </div>
          </div>
        </div>
        <div class="shop-cart__content">
          <ShopCartItem
            v-for="(item, index) in cartItems"
            :key="getShopCartItemKey(item.id || index)"
            :cart-item="item"
            @increase-count="increaseCountItem"
            @delete-item="deleteItem"
          />
        </div>
        <div class="shop-cart__sum">
          <span class="shop-cart__sum-text">
            Общая стоимость: <span class="shop-cart__sum-text-price">{{ userCartSum }} руб.</span>
          </span>
        </div>
      </div>
    </div>
  </section>
</template>

<style lang="scss">
@import "../../styles/components/shop-cart/shop-cart";
</style>
