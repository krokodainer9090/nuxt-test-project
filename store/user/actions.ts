// Catalog Types
import { ShopCatalogClient, ShopCatalogGroup } from '~/store/shop-catalog/types';

// User Types
import { UserState, ShopItem } from '~/store/user/types';

// Vuex
import { ActionContext, ActionTree } from 'vuex';

const actions: ActionTree<UserState, any> = {
  addItemToCart ({ commit, dispatch }: ActionContext<UserState, any>, item: ShopItem) {
    commit('ADD_TO_USER_CART', item);
  },

  increaseItemCount ({ commit, dispatch }: ActionContext<UserState, any>, item: ShopItem) {
    commit('INCREASE_USER_CART_ITEM', item);
  },

  removeItemFromCart ({ commit, dispatch }: ActionContext<UserState, any>, itemId: number) {
    commit('REMOVE_FROM_USER_CART', itemId);
  },

  updateCartPrices ({ commit }: ActionContext<UserState, any>, catalogData: ShopCatalogClient) {
    const catalogItems = catalogData.reduce((acum: any[], group: ShopCatalogGroup) => {
      if ( group.items ) {
        acum.push(...group.items);
      }

      return acum;
    }, []);

    commit('UPDATE_CART_PRICES', catalogItems);
  }
};

export default actions;
