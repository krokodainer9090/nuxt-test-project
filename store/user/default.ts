import { UserCart, UserCurrency } from '~/store/user/types';

export const defaultUserCard: UserCart = [];
export const defaultCurrencyValue: UserCurrency = 20;
