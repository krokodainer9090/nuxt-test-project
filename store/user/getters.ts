import { UserCart, UserCurrency, UserState } from '~/store/user/types';

export default {
  getUserCart (state: UserState): UserCart {
    return state.userCart;
  },

  getUserCartSum (state: UserState): number {
    const sum = state.userCart.reduce((accum, currentValue) => {
      accum += currentValue.price * currentValue.currentCount;

      return accum;
    }, 0);

    return Math.round(sum * 100) / 100;
  },

  getUserCurrencyValue (state: UserState): UserCurrency {
    return state.currencyValue;
  }
};
