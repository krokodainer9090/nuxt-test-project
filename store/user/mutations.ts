// Catalog Types
import { ShopCatalogItem } from '~/store/shop-catalog/types';

// User Types
import { UserState, ShopItem, UserCurrency } from '~/store/user/types';

// Vuex
import { MutationTree } from 'vuex';

const mutations: MutationTree<UserState> = {
  ADD_TO_USER_CART (state: UserState, item: ShopItem) {
    const isAlreadyExits = state.userCart.find(cartItem => {
      return cartItem.id === item.id;
    });

    if ( !isAlreadyExits ) {
      item = {
        ...item,
        currentCount: 1
      };

      state.userCart.push(item);
    }
  },

  INCREASE_USER_CART_ITEM (state: UserState, { id, count }: ShopItem) {
    const itemIndex = state.userCart.findIndex(cartItem => {
      return cartItem.id === id;
    });

    if ( itemIndex !== -1 ) {
      state.userCart[itemIndex].currentCount = Number(count);
    }
  },

  REMOVE_FROM_USER_CART (state: UserState, itemId: number) {
    const deleteItemIndex = state.userCart.findIndex(cartItem => {
      return cartItem.id === itemId;
    });

    if ( deleteItemIndex !== -1 ) {
      state.userCart.splice(deleteItemIndex, 1);
    }
  },

  SET_CURRENCY_VALUE (state: UserState, currencyValue: UserCurrency) {
    state.currencyValue = currencyValue;
  },

  UPDATE_CART_PRICES (state: UserState, catalogItems: ShopCatalogItem[]) {
    state.userCart.forEach((item, key) => {
      const itemInCatalog = catalogItems.find((catalogItem: ShopCatalogItem) => catalogItem.id === item.id);

      if ( itemInCatalog ) {
        state.userCart[key].price = itemInCatalog.price;
      }
    });
  }
};

export default mutations;
