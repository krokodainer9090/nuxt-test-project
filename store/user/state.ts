// Types
import { UserState } from '~/store/user/types';

// Default
import { defaultUserCard, defaultCurrencyValue } from '~/store/user/default';

export default (): UserState => {
  return {
    userCart: defaultUserCard,
    currencyValue: defaultCurrencyValue
  };
};
