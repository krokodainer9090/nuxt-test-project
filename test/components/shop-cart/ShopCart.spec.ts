// Components
import ShopCart from '~/components/shop-cart/ShopCart.vue';
import ShopCartItem from '~/components/shop-cart/ShopCartItem.vue';

// Utilities
import { shallowMount, Wrapper, createLocalVue } from '@vue/test-utils';
import { shopCatalogItem } from '~/test/utils/misc/constants';

// Store
import { userStore } from '~/test/store/user/user-store';

const localVue = createLocalVue();
const store = userStore(localVue);

// Constants
const newCountItem = 10;

describe('ShopCart', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(ShopCart, {
      ...options,
      localVue,
      store
    });
  };

  it('should render empty cart', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render cart with item', async () => {
    await store.dispatch('user/addItemToCart', shopCatalogItem);

    const wrapper = mountFunction();
    const itemComponent = wrapper.findComponent(ShopCartItem);

    expect(wrapper.html()).toMatchSnapshot();

    // Increase count of item
    itemComponent.vm.$emit('increase-count', {
      id: shopCatalogItem.id,
      count: newCountItem
    });
    shopCatalogItem.currentCount = newCountItem;

    expect(wrapper.vm.cartItems).toEqual([shopCatalogItem]);

    // Delete item from cart
    itemComponent.vm.$emit('delete-item', shopCatalogItem.id);

    expect(wrapper.vm.cartItems).toEqual([]);
  });
});
