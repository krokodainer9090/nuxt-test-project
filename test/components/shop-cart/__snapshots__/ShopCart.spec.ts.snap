// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCart should render cart with item 1`] = `
"<section class=\\"shop-cart\\">
  <div class=\\"container\\">
    <div class=\\"shop-cart__content\\">
      <h2 class=\\"shop-cart__title\\">Корзина</h2>
      <div class=\\"shop-cart__header\\">
        <div class=\\"row\\">
          <div class=\\"col-6\\"><span class=\\"shop-cart__header-text\\">
              Наименование товара и описание
            </span></div>
          <div class=\\"col-2\\"><span class=\\"shop-cart__header-text\\">
              Количество
            </span></div>
          <div class=\\"col-4\\"><span class=\\"shop-cart__header-text\\">
              Цена
            </span></div>
        </div>
      </div>
      <div class=\\"shop-cart__content\\">
        <shopcartitem-stub cartitem=\\"[object Object]\\"></shopcartitem-stub>
      </div>
      <div class=\\"shop-cart__sum\\"><span class=\\"shop-cart__sum-text\\">
          Общая стоимость: <span class=\\"shop-cart__sum-text-price\\">42.8 руб.</span></span></div>
    </div>
  </div>
</section>"
`;

exports[`ShopCart should render empty cart 1`] = `
"<section class=\\"shop-cart\\">
  <div class=\\"container\\">
    <div class=\\"shop-cart__content\\">
      <h2 class=\\"shop-cart__title\\">Корзина</h2>
      <div class=\\"shop-cart__header\\">
        <div class=\\"row\\">
          <div class=\\"col-6\\"><span class=\\"shop-cart__header-text\\">
              Наименование товара и описание
            </span></div>
          <div class=\\"col-2\\"><span class=\\"shop-cart__header-text\\">
              Количество
            </span></div>
          <div class=\\"col-4\\"><span class=\\"shop-cart__header-text\\">
              Цена
            </span></div>
        </div>
      </div>
      <div class=\\"shop-cart__content\\"></div>
      <!---->
    </div>
  </div>
</section>"
`;
